sql-plugin/src/main/scala/org/apache/spark/sql/rapids/execution/python/GpuWindowInPandasExec.scala

```diff
@@ -0,0 +1,570 @@
+        + "partition, this can cause serious performance degradation.")
+      AllTuples :: Nil
+    } else {
+      ClusteredDistribution(partitionSpec) :: Nil
+    }
+  }
+
+  override def requiredChildOrdering: Seq[Seq[SortOrder]] =
+    Seq(partitionSpec.map(SortOrder(_, Ascending)) ++ orderSpec)
+
+  override def outputOrdering: Seq[SortOrder] = child.outputOrdering
+
+  override def outputPartitioning: Partitioning = child.outputPartitioning
+
+  override def childrenCoalesceGoal: Seq[CoalesceGoal] = Seq(RequireSingleBatch)
+
+  /*
+   * Helper functions and data structures for window bounds
+   *
+   * It contains:
+   * (1) Function from frame index to its lower bound column index in the python input row
+   * (2) Function from frame index to its upper bound column index in the python input row
+   * (3) Seq from frame index to its window bound type
+   */
+  private type WindowBoundHelpers = (Int => Int, Int => Int, Seq[WindowBoundType])
+
+  /*
+   * Enum for window bound types. Used only inside this class.
+   */
+  private sealed case class WindowBoundType(value: String)
+  private object UnboundedWindow extends WindowBoundType("unbounded")
+  private object BoundedWindow extends WindowBoundType("bounded")
+
+  private val windowBoundTypeConf = "pandas_window_bound_types"
+
+  private def collectFunctions(udf: GpuPythonUDF): (ChainedPythonFunctions, Seq[Expression]) = {
+    udf.children match {
+      case Seq(u: GpuPythonUDF) =>
+        val (chained, children) = collectFunctions(u)
+        (ChainedPythonFunctions(chained.funcs ++ Seq(udf.func)), children)
+      case children =>
+        // There should not be any other UDFs, or the children can't be evaluated directly.
+        assert(children.forall(_.find(_.isInstanceOf[GpuPythonUDF]).isEmpty))
+        (ChainedPythonFunctions(Seq(udf.func)), udf.children)
+    }
+  }
+
+  // Similar with WindowExecBase.windowFrameExpressionFactoryPairs
+  // but the functions are not needed here.
+  private lazy val windowFramesWithExpressions = {
+    type FrameKey = (String, GpuSpecifiedWindowFrame)
+    type ExpressionBuffer = mutable.Buffer[Expression]
+    val framedExpressions = mutable.Map.empty[FrameKey, ExpressionBuffer]
+
+    // Add expressions to the map for a given frame.
+    def collect(tpe: String, fr: GpuSpecifiedWindowFrame, e: Expression): Unit = {
+      val key = (tpe, fr)
+      val es = framedExpressions.getOrElseUpdate(key, mutable.ArrayBuffer.empty[Expression])
+      es += e
+    }
+
+    // Collect all valid window expressions and group them by their frame.
+    windowExpression.foreach { e =>
+      e.foreach {
+        case e @ GpuWindowExpression(function, spec) =>
+          val frame = spec.frameSpecification.asInstanceOf[GpuSpecifiedWindowFrame]
+          function match {
+            case GpuAggregateExpression(_, _, _, _, _) => collect("AGGREGATE", frame, e)
+            case _: GpuAggregateWindowFunction => collect("AGGREGATE", frame, e)
```

Comment:
Do we actually support these functions here?

Reply:
For CPU version, this logic is also used by `WindowExec` besides `WindowInPandasExec`.
For GPU version currently, is it only used by `GpuWindowInPandasExec `.
We may support it in our `GpuWindowExec` (@mythrocks pls correct me if wrong), but not in `GpuWindowInPandasExec `, so i think it is ok to remove this if you prefer.

Reply:
OK So I read the original Spark code and this is not anywhere in that code. They do not support anything but a PythonUDF

https://github.com/apache/spark/blob/1eb236b9360a000afc30424341698fe26ee96d0f/sql/core/src/main/scala/org/apache/spark/sql/execution/python/WindowInPandasExec.scala#L206

We should do the same and delete all of the extra code, and possibly throw an exception if we see anything else.

Reply:
Yes, we have this too, it is here https://github.com/NVIDIA/spark-rapids/blob/d865883ee35034f119e4e014ad1e89d354688907/sql-plugin/src/main/scala/org/apache/spark/sql/rapids/execution/python/GpuWindowInPandasExec.scala#L422 

The code about collecting the functions is similar with `windowFrameExpressionFactoryPairs ` in  `WindowExecBase` https://github.com/apache/spark/blob/master/sql/core/src/main/scala/org/apache/spark/sql/execution/window/WindowExecBase.scala#L118 ,
which is used to filter the Python UDF as you mentioned.

Reply:
I move this logic into the `GpuWindowInPandasExec`, then our implementation does not need to extends from `WindowExecBase` any more.